### ISSUE TITLE
Detect if a field is not set by the sender

### ISSUE THREAD

**Issue by maximg** [maximg/thrift#29]
It is possible that not all fields are set by the sender; the receiver would like to know which fields were not received. C++ implementation uses a bit array for that; there may be better ways in Rust (like Option).

**Comment by gsingh93** [comment]
Can you clarify what this means? Are you talking about `optional` fields in the thrift specification or something else?

**Comment by maximg** [comment]
Yes, this is about proper support for `required` and `optional` fields.

C++ and Java provide setters for struct fields and use __isset bit flags to detect if a field was set; these flags are also populated by the receiver to indicate which fields were received from the sender.

Some details on these attributes are available here: http://diwakergupta.github.io/thrift-missing-guide/#_defining_structs and in this JIRA issue: https://issues.apache.org/jira/browse/THRIFT-1394

Here is a possible way to model fields with different attributes:

`required` (`T_REQUIRED`): `Option<type>`; always sent. Sender shall complain if the field is not `Some(v)`. Receiver raises a format violation if the field is not received.

`optional` without default (`T_OPTIONAL`): `Option<type>`. Sent if `Some()`. No further checks by the sender or the receiver. If not received, receiver populates with `None`.

`optional` with default (`T_OPTIONAL`): `Option<type>`; Initialized to `Some(default)`. Sent if `Some(v) && v != default` (can this get expensive for complex structures?).  If not received, receiver populates with `Some(default)`. The default value is never sent over the wire.

no attributes (T_OPT_IN_REQ_OUT): `Option<type>`; always sent. Sender shall complain if the field is not `Some(v)` (REQ_OUT). If not received, receiver populates with `None` (OPT_IN).

This issue seems to duplicate #10 so I am closing that one and restating the title of this issue.

It may be worth checking if isset flags are ever sent over the wire - I could not confirm it yet.

**Comment by maximg** [comment]
It seems these rules only apply to primitive fields and structs; the containers can/should be handled differently, as for a container logically there may be no difference between a "not present" and "empty" state (does this agree with the way other implementations treat them?). So we could model containers as just as fields of `type`.

@sgnr, @gsingh93 - as the changes in this issue affect the client API, could you review the model suggested? Do you see better alternatives?
